### PR TITLE
Adding basic support for gitlab

### PR DESCRIPTION
**Pull request by jangroth**
Thanks for implementing Keydra!

Here at [Adatree](https://www.adatree.com.au), we were looking for a tool almost extactly like Keydra, only that we require support for GitLab as a client/provider.

We have decided to use Keydra as a base and added our own implementation to support GitLab. This has been working great for us, we are using Keydra in production for a while now.

If you are interested - we'd love to contribute back to the core project.

This PR contains a GitLab client/provider, similar to the existing GitHub client/provider:
* Support for distribution only (rotation not implemented)
* Utilizing GitLab's Python client (`python-gitlab`)
* Using `schema` for validation of configuration
  * See `keydra/providers/gitlab/SPEC_SCHEMA`
  * Note: This could be interesting for other aspects of Kedra too, e.g. `keydra/providers/base/validate_spec()`
* New configuration feature:
  * Specify git branch for secrets/envs files (useful for development)
* The shady bits:
  * I struggled to fully grasp the intention of spec-pre-processing
    * GitLab provider & tests do what GitHub provider & tests are doing, but this is very much 'as is' only
  * No documentation so far
    * Waiting for initial feedback from the maintainers, happy to add afterwards

Please let us know what you think.

Kind regards & thanks again for implementing such a useful tool!

### CODE REVIEW

File: src/keydra/clients/gitlab.py
Comment (vludax) [inline src/keydra/clients/gitlab.py]: Could be useful to log `repo_name` too :)

File: src/keydra/providers/gitlab.py
Comment (vludax) [inline src/keydra/providers/gitlab.py]: This may be a little misleading -- the distributed value is not always `key`, but rather `dest['source']`, right?

File: src/keydra/providers/gitlab.py
Comment (vludax) [inline src/keydra/providers/gitlab.py]: Any reason for not using `LOGGER.warn` here?

**Review by vludax — COMMENTED** [review]
Hi jangroth,

Thank you for your contribution and feedback!

I'm sure the GitLab provider will be useful for other teams, and will enable broader adoption of Keydra. The code looks good, with only a few minor nit-picks below. The fact that you've already used it in production is obviously a good indicator of it working!

I'm quite fond of the idea of using declarative schema-based spec validation rather than the procedural approach we've taken in other providers. We will probably look at switching them over at some stage, e.g. when we need to make changes to them.

It may be worth sharing another idea we had regarding spec validation -- we wanted to enable this validation via pull request checks in the Keydra config repo, as opposed to the run-time validation in Keydra itself. We haven't settled on the preferred way to execute this yet -- ideas welcome 🙂

I must admit that the spec pre-processing piece is a little confusing. One (only?) purpose I see it serving is setting the environment name dynamically during distribution, e.g.:

```
- key: SOME_PASSWORD
  provider: bitbucket
  source: secret_field_name
  config:
    scope: deployment
    create: true
    environment: '{ENV}'
    repository:
      - target_repo
  envs:
    - dev
    - test
    - prod
```

In one use case we use the `{ENV}` var to populate the name of Bitbucket variables, i.e. `key`, although we intend to deprecate this from our config:
```
key: '{ENV}_SF_USERNAME'
provider: bitbucket
... 
```

I'll let @gyrospectre add anything I've missed 🙂

**Comment by jangroth** [comment]
Thanks @vludax  and @gyrospectre , appreciate your friendly response.

I won't be able to work on this before the end of the week, so any more feedback till then just throw it in... 😄

File: src/keydra/providers/gitlab.py
Comment (jangroth) [reply src/keydra/providers/gitlab.py]: I kept the log message identical with its counterpart from the github provider (https://github.com/Athena-Home-Loans/keydra/blob/9bab1b21e025ceb6ae074ea936d693e36efae5a4/src/keydra/providers/github.py#L58)

Happy change of course, do you have a suggestion?

File: src/keydra/providers/gitlab.py
Comment (jangroth) [reply src/keydra/providers/gitlab.py]: Oh, nice catch.

**Comment by jangroth** [comment]
Hey @vludax  and @gyrospectre , I've implement a few - tiny - code changes based on your suggestions. I've also added documentation for the gitlab bits that we've added. Finally, I'm removing draft status from the PR.

> It may be worth sharing another idea we had regarding spec validation -- we wanted to enable this validation via pull request checks in the Keydra config repo, as opposed to the run-time validation in Keydra itself.

That's an interesting idea. Clearly you guys must be Keydra at a larger scale then we are :-) I think it's a great addition to make things easier to use/fail faster. However, it also requires something to be installed somewhere, right? (I assume you are thinking server-side commit hooks in Bitbucket?). Again, I think it's a great idea, but it'll be good to avoid locking Keydra into proprietary products. Maybe a Makefile target could be a good non-proprietary comprise, e.g. `make check-config`. Obviously very easy to execute, but also reasonably easy to integrate into different git providers (BB, GL, GH). This could also go nicely into (client-side) commit hooks... 🤔

**Comment by gyrospectre** [comment]
Looks good, thanks for your contribution @jangroth! Will wait for @vludax to review as the majority of comments were from him, then will approve if all good.

File: docs/_sources/providers.rst.txt
Comment (vludax) [inline docs/_sources/providers.rst.txt]: Assume we mean "Gitlab" 🙂

File: docs/_sources/providers.rst.txt
Comment (vludax) [inline docs/_sources/providers.rst.txt]: `gitlab`?

File: src/keydra/providers/gitlab.py
Comment (vludax) [reply src/keydra/providers/gitlab.py]: `'Successfully distributed {}.{}.{} to '` with `dest['source']` added as a parameter to `.format()` should cut it 🙂

**Comment by vludax** [comment]
Happy with the current state overall, just a few minor copy-paste issues.

> That's an interesting idea. Clearly you guys must be Keydra at a larger scale then we are :-) I think it's a great addition to make things easier to use/fail faster. However, it also requires something to be installed somewhere, right? (I assume you are thinking server-side commit hooks in Bitbucket?). Again, I think it's a great idea, but it'll be good to avoid locking Keydra into proprietary products. Maybe a Makefile target could be a good non-proprietary comprise, e.g. `make check-config`. Obviously very easy to execute, but also reasonably easy to integrate into different git providers (BB, GL, GH). This could also go nicely into (client-side) commit hooks... 🤔

Yep, definitely wouldn't want to couple this to any kind of CI/CD environment. As a trivial solution, we could indeed checkout `keydra` repo and run a CLI command to perform config validation. Another option could be hosting it as a separate Lambda and calling it from pipelines. A perhaps more appropriate alternative could be publishing `keydra-validator` to https://pypi.org/, which can be installed from pipelines or executed locally. This also reminds of the point we discussed earlier -- publishing Keydra itself as a package that can be deployed anywhere. This would need some refactoring to abstract away the heavy AWS integration we currently have in Keydra.

**Comment by jangroth** [comment]
>  This also reminds of the point we discussed earlier -- publishing Keydra itself as a package that can be deployed anywhere.

That thought has definitely crossed my mind too 😃  It would certainly make everyone's life a little easier, but I agree it's not a small refactoring. 

As an intermediate step, you could maybe consider making Keydra pip-installable, without going all the way to `pypi`. I remember working on similar projects where we needed an easy way to get the code into our pipelines - something like `pip install git+ssh://git@github.com/the-project` did the trick for us.

**Review by vludax — APPROVED** [review]
Thanks for your contribution jangroth!

**Pull request merged**
